inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/InstrumentationManager.java

```diff
@@ -0,0 +1,209 @@
+            CacheBuilder.newBuilder().weakKeys().<Class<?>, Boolean>build().asMap());
+
+    private BatchJobExecutorService.BatchJob<BatchSize> classInstrumentationJob;
+
+    @Value
+    private static class BatchSize {
+        private int maxClassesToCheck;
+        private int maxClassesToRetransform;
+    }
+
+
+    @PostConstruct
+    private void init() {
+        InternalSettings conf = env.getCurrentConfig().getInstrumentation().getInternal();
+        val batchSizes = new BatchSize(conf.getClassConfigurationCheckBatchSize(), conf.getClassRetransformBatchSize());
+        Duration delay = conf.getInterBatchDelay();
+
+        classInstrumentationJob = executor.startJob(this::checkClassesForConfigurationUpdates, batchSizes, delay, delay);
+    }
+
+    @PreDestroy
+    private void destroy() {
+        classInstrumentationJob.cancel();
+    }
+
+    @Override
+    public void newClassesDiscovered(Set<Class<?>> newClasses) {
+        loadedClasses.addAll(newClasses);
+        pendingClasses.addAll(newClasses);
+    }
+
+    @EventListener
+    private void classInstrumented(ClassInstrumentedEvent event) {
+        ClassInstrumentationConfiguration config = event.getAppliedConfiguration();
+        TypeDescription type = event.getInstrumentedClassDescription();
+        Class<?> clazz = event.getInstrumentedClass();
+
+        if (ClassInstrumentationConfiguration.NO_INSTRUMENTATION.isSameAs(type, config)) {
+            activeInstrumentations.remove(clazz);
+        } else {
+            activeInstrumentations.put(clazz, config);
+        }
+
+        //there might be race conditions when a class is being retransformed
+        //and we considered it up-to-date at the same time
+        //for this reason we have to recheck every class after it has been instrumented
+        pendingClasses.add(clazz);
+    }
+
+    @EventListener
+    private void configEventListener(InspectitConfigChangedEvent ev) {
+
+        InternalSettings newInternal = ev.getNewConfig().getInstrumentation().getInternal();
+        val batchSizes = new BatchSize(newInternal.getClassConfigurationCheckBatchSize(), newInternal.getClassRetransformBatchSize());
+        classInstrumentationJob.setBatchSizes(batchSizes);
+        classInstrumentationJob.setInterBatchDelay(newInternal.getInterBatchDelay());
+    }
+
+    @EventListener
+    private void instrumentationConfigEventListener(InstrumentationConfigurationChangedEvent ev) {
+        pendingClasses.addAll(loadedClasses);
+    }
+
+
+    /**
+     * Processes a given amount of classes from {@link #pendingClasses}.
+     * For the classes where it is required a retransform is triggered.
+     * <p>
+     * This method is package-private for testing.
+     *
+     * @param batchSize the number of classes to take from {@link #pendingClasses} and to retransform per batch
+     */
+    void checkClassesForConfigurationUpdates(BatchSize batchSize) {
+        if (transformer.isShuttingDown()) {
```

Comment:
the only reason we are auto-wiring the `transformer` is because of the shut down.. I see we also keep track of the shut-down in the `NewClassDiscoveryService`.. Can you quickly check if Spring offers some helper for context shutting down that we would use?